src/main/java/frc/robot/subsystems/internals/InternalSubsystem.java

```diff
@@ -92,79 +99,60 @@ public InternalSubsystem(TurretSubsystem turretSubsystem) {
+        if (!prevEntranceDetected && entranceDetected) entranceStorageBallCount++;
         prevEntranceDetected = entranceDetected;
 
-        // If a ball has entered internals, start the bottom motor
-        if (entranceDetected) motorBottom.set(0.3);
-
-        // If there is a ball in storage, start the top motor
-        if (isBall(storageColor)) {
+        Color storageColor = matchColor(colorSensorThread.getLastStorage());
+        if (!prevStorageDetected && isBall(storageColor)) {
             // If we haven't already checked rejection logic, reject the ball if it doesn't match alliance color
             if (!rejectingChecked) {
                 rejecting = storageColor != ALLIANCE_COLOR;
                 rejectingChecked = true;
             }
-            motorTop.set(0.5);
-            //System.out.println("STORAGE detected, running top motor");
+            entranceStorageBallCount--;
+            storageStagingBallCount++;
         }
 
-        // If there is a ball in staging, stop the bottom and top motors
-        if (stagingDetected) {
-            motorBottom.set(0);
-            motorTop.set(0);
-            //System.out.println("STAGING detected, stopping both motors");
+        boolean stagingDetected = staging.get() >= 0.2;
+        if (!prevStagingDetected && stagingDetected) {
+            storageStagingBallCount--;
+            stagingExitBallCount++;
         }
+        prevStagingDetected = stagingDetected;
+
+        if (driverOverride) return;
+
+        // If there is a ball in the entrance or between storage and staging *and* staging is empty, run the bottom motor.
+        // This serves to bring the first ball properly from storage to staging. For the second ball, the right hand side
+        // of the condition will evaluate to false so it will be stopped when the ball reaches storage.
+        motorBottom.set(entranceStorageBallCount > 0 || (stagingExitBallCount == 0 && storageStagingBallCount > 0)
+            ? 0.3 : 0);
+
+        // If there is a ball between storage and staging, run the top motor
+        motorTop.set(storageStagingBallCount > 0 ? 0.5 : 0);
 
         if (turretSubsystem != null) {
             // If a shot was requested and the turret is ready, load a ball into the turret.
             // If rejecting, the turret can be in a semi-aligned state; otherwise, require it to be fully lined up.
             turretSubsystem.setReject(rejecting);
-            if (shotRequested /* && turretSubsystem.getState() == TurretSubsystem.ModuleState.READY
-                 || rejecting && turretSubsystem.getState() == TurretSubsystem.ModuleState.ALMOST */) {
+            if (shotRequested /* && turretSubsystem.getState() == TurretSubsystem.ModuleState.HIGH_TOLERANCE
+                 || rejecting && turretSubsystem.getState() == TurretSubsystem.ModuleState.LOW_TOLERANCE */) {
                 // Spin the top motor on a timer
                 exitTimer.start();
                 motorTop.set(0.5);
```

Comment:
won't this and line 134 potentially cause jittering? might be best to use booleans till the very end